src/compiler/binder.ts

```diff
@@ -3049,7 +3049,11 @@ namespace ts {
         function bindSpecialPropertyAssignment(node: BindablePropertyAssignmentExpression) {
             // Class declarations in Typescript do not allow property declarations
             const parentSymbol = lookupSymbolForPropertyAccess(node.left.expression, container) || lookupSymbolForPropertyAccess(node.left.expression, blockScopeContainer) ;
-            if (!isInJSFile(node) && !isFunctionSymbol(parentSymbol)) {
+            const isFunctionSym = isFunctionSymbol(parentSymbol);
+            if (!isInJSFile(node) && !isFunctionSym) {
+                return;
+            }
+            if (isFunctionSym && getDeclarationName(cast(node.left, isBindableStaticNameExpression)) === "length") {

```

Comment:
I'm not super happy about this fix but `binder` doesn't have access any access to the type information so I can't just check what is declared on the `interface Function` as readonly.

And I think that it's better to avoid binding this property altogether than trying to account for this case in the `checker` or to write some fixup logic in between the binder and the checker.

I've originally described my conclusion after the debugging session here: https://github.com/microsoft/TypeScript/issues/49113#issuecomment-1128594794